omnigibson/utils/usd_utils.py

```diff
@@ -850,9 +850,32 @@ def __init__(self, pattern):
         # Mapping from prim path to base footprint link name if one exists, None if the root is the base link.
         self._base_footprint_link_names = {}
 
-    def clear(self):
+        # Prior link transforms / dof positions for estimating velocities since Omni gives inaccurate values
+        self._last_state = None
+
+    def post_physics_step(self):
+        # Should be called every sim physics step, right after a new physics step occurs
+        # The current poses (if it exists) are now the former poses from the previous timestep
+        if "root_transforms" in self._read_cache:
+            self._last_state = {
+                "root_transforms": cb.copy(self._read_cache["root_transforms"]),
+                "link_transforms": cb.copy(self._read_cache["link_transforms"]),
+                "dof_positions": cb.copy(self._read_cache["dof_positions"]),
+            }
+
+        # Clear the internal data since everything is outdated
+        self.clear(keep_last_pose=True)
+
+    def clear(self, keep_last_pose=False):
         self._read_cache = {}
         self._write_idx_cache = collections.defaultdict(set)
+        if not keep_last_pose:
+            self._last_state = None
+
+        # Update the transforms internally so that they're guaranteed to exist during the beginning of the next timestep
+        self._read_cache["root_transforms"] = cb.from_torch(self._view.get_root_transforms())
+        self._read_cache["link_transforms"] = cb.from_torch(self._view.get_link_transforms())
+        self._read_cache["dof_positions"] = cb.from_torch(self._view.get_dof_positions())
 
```

Comment:
I am quite confused with the logic here actually. Can we discuss in person?

Reply:
resolved in person